pennylane/devices/preprocess.py

```diff
@@ -147,6 +147,20 @@ def validate_device_wires(
     return (tape,), null_postprocessing
 
 
+@transform
+def mid_circuit_measurements(
+    tape: qml.tape.QuantumTape, device
+) -> (Sequence[qml.tape.QuantumTape], Callable):
+    """Provide the transform to handle mid-circuit measurements.
+
+    If the tape of device uses finite-shot, use the native implementation (i.e. no transform), and use defer measurements transforms otherwise.
+    """
+    has_shots = tape.shots.total_shots is not None or device.shots.total_shots
```

Comment:
You can strictly pull the number of shots from the tape.  `device.shots` is only used to initialize the number of shots on the tape at the start of the workflow. `Device.execute` should not read from `Device.shots`, only `circuit.shots`.

You can also treat `tape.shots` as a boolean (`if tape.shots`).

Reply:
Thanks for the advice, I'll make the changes.